File: haskell/Exercises.hs
module Exercises
  ( change,
    firstThenApply,
    powers,
    meaningfulLineCount,
    Shape (Box, Sphere),
    volume,
    surfaceArea,
    BST (Empty),
    size,
    insert,
    contains,
    inorder,
  )
where

import Data.Char (isSpace)
import Data.List (find, isPrefixOf)
import Data.Map qualified as Map
import Data.Text (pack, replace, unpack)

change :: Integer -> Either String (Map.Map Integer Integer)
change amount
  | amount < 0 = Left "amount cannot be negative"
  | otherwise = Right $ changeHelper [25, 10, 5, 1] amount Map.empty
  where
    changeHelper [] remaining counts = counts
    changeHelper (d : ds) remaining counts =
      changeHelper ds newRemaining newCounts
      where
        (count, newRemaining) = remaining `divMod` d
        newCounts = Map.insert d count counts

firstThenApply :: [a] -> (a -> Bool) -> (a -> b) -> Maybe b
firstThenApply xs pred f = fmap f (find pred xs)

powers :: Integral a => a -> [a]
powers base = map (base ^) [0 ..]

meaningfulLineCount :: FilePath -> IO Int
meaningfulLineCount path = do
  contents <- readFile path
  return $ length $ filter meaningfulLine $ lines contents
  where
    meaningfulLine line =
      let trimmedLine = dropWhile isSpace line
       in not (null trimmedLine) && not ("#" `isPrefixOf` trimmedLine)

data Shape
  = Sphere Double
  | Box Double Double Double
  deriving (Eq, Show)

volume :: Shape -> Double
volume (Sphere r) = (4 / 3) * pi * r ^ 3
volume (Box l w h) = l * w * h

surfaceArea :: Shape -> Double
surfaceArea (Sphere r) = 4 * pi * r ^ 2
surfaceArea (Box l w h) = 2 * (w * l + w * h + l * h)

data BST a
  = Empty
  | Node a (BST a) (BST a)

size :: BST a -> Int
size Empty = 0
size (Node _ left right) = 1 + size left + size right

inorder :: BST a -> [a]
inorder Empty = []
inorder (Node value left right) = inorder left ++ [value] ++ inorder right

insert :: (Ord a) => a -> BST a -> BST a
insert value Empty = Node value Empty Empty
insert value (Node nodeValue left right)
  | value < nodeValue = Node nodeValue (insert value left) right
  | value > nodeValue = Node nodeValue left (insert value right)
  | otherwise = Node nodeValue left right

contains :: Eq a => a -> BST a -> Bool
contains _ Empty = False
contains x (Node value left right) = x == value || contains x left || contains x right

instance (Show a) => Show (BST a) where
  show :: (Show a) => BST a -> String
  show Empty = "()"
  show tree = showNode tree
    where
      showNode :: (Show a) => BST a -> String
      showNode Empty = ""
      showNode (Node value left right) = "(" ++ showNode left ++ show value ++ showNode right ++ ")"

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

let first_then_apply array predicate consumer =
  match List.find_opt predicate array with
  | None -> None
  | Some x -> consumer x

let powers_generator base =
  let rec generate_from power () =
    Seq.Cons (power, generate_from (base * power))
  in
  generate_from 1

let meaningful_line_count filename = 
  let meaningful_line line = 
    let trimmed = String.trim line in
    String.length trimmed > 0 && not (String.get trimmed 0 = '#')
  in

  let the_file = open_in filename in
  let finally () = close_in the_file in

  let rec count_lines count =
    try
      let line = input_line the_file in
      if meaningful_line line then
        count_lines (count + 1)
      else
        count_lines (count)
    with
      End_of_file -> count
    in

  Fun.protect ~finally (fun () ->  count_lines 0)


type shape =
  | Sphere of float
  | Box of float * float * float

let volume s =
  match s with
  | Sphere r -> Float.pi *. (r ** 3.) *. 4. /. 3.
  | Box (l, w, h) -> l *. w *. h

let surface_area s =
  match s with
  | Sphere r -> Float.pi *. 4. *. (r ** 2.)
  | Box (l, w, h) -> ((w *. l) +. (w *. h) +. (l *. h)) *. 2.


type 'a binary_search_tree =
  | Empty
  | Node of 'a binary_search_tree * 'a * 'a binary_search_tree

let rec size tree = 
  match tree with
  | Empty -> 0
  | Node (left, _, right) -> 1 + size left + size right

let rec contains value tree = 
  match tree with
  | Empty -> false
  | Node (left, v, right) -> 
    if v = value then true
    else if value < v then contains value left
    else contains value right

let rec insert value tree = 
  match tree with
  | Empty -> Node (Empty, value, Empty)
  | Node (left, v, right) ->
    if value < v then
      Node (insert value left, v, right)
    else if value > v then
      Node (left, v, insert value right)
    else
      Node (left, v, right)

let rec inorder tree = 
  match tree with
  | Empty -> []
  | Node (left, v, right) -> inorder left @ [v] @ inorder right
